test/k23si/3SITxnTest.cpp

```diff
@@ -1794,10 +1797,690 @@ seastar::future<> testScenario02() {
 seastar::future<> testScenario03() {
     return seastar::make_ready_future()
     .then([] {
-        return seastar::sleep(1s);
+        return seastar::sleep(300ms);
     });
 }
 
+seastar::future<> testScenario04() {
+    std::cout << std::endl << std::endl;
+	K2INFO("+++++++ TestScenario 04: read your writes and read-committed isolation +++++++");
+	K2INFO("--->Test SETUP: initialization record, pKey(\"SC04_pkey1\"), rKey(\"rKey1\"), v0{f1=SC04_f1_zero, f2=SC04_f2_zero} -> committed");
+    
+    return seastar::make_ready_future()
+    .then([] {
+        return getTimeNow();
+    })
+    .then([this](dto::Timestamp&& ts) {
+        return seastar::do_with(
+            dto::K23SI_MTR {.txnid = txnids++, .timestamp = std::move(ts), .priority = dto::TxnPriority::Medium},
+            dto::Key {.schemaName = "schema", .partitionKey = "SC04_pkey1", .rangeKey = "rKey1"},
+            dto::Key {.schemaName = "schema", .partitionKey = "SC04_pkey1", .rangeKey = "rKey1"},
+            DataRec {.f1="SC04_f1_zero", .f2="SC04_f2_zero"},
+            [this](auto& mtr, auto& k1, auto& trh, auto& v0) {
+                return doWrite(k1, v0, mtr, trh, collname, false, true, ErrorCaseOpt::NoInjection)
+                .then([](auto&& response) {
+                    auto& [status, resp] = response;
+                    K2EXPECT(status, dto::K23SIStatus::Created);
+                })
+                .then([this, &trh, &mtr, &k1] {
+                    return doEnd(trh, mtr, collname, true, {k1}, Duration(0s), ErrorCaseOpt::NoInjection)
+                    .then([](auto&& response) {
+                        auto& [status, resp] = response;
+                        K2EXPECT(status, dto::K23SIStatus::OK);
+                    });
+                });
+            }
+        );
+    })
+    .then([]{
+        K2INFO("Scenario 04 setup done.");
+        return getTimeNow();
+    })
+    .then([this](dto::Timestamp&& ts) {
+        std::cout << std::endl;
+
+        return seastar::do_with(
```

Comment:
These more complicated scenarios are getting a bit hard to follow in the code, especially with the nesting. Maybe you can refactor the scenario into a class and use member variables instead of do_with clauses?

Reply:
A test logic figure for  scenario 04 is added. And I refactor the implementation to make it more easier to read. To keep the style consistent with the rest of the scenarios, I did not refactor it into a class and member variables.